### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by Ashrom0**

### CODE REVIEW

**Comment by TechGlue** [comment]
I will review this!

File: SecretSanta/src/SecretSanta.Api/Controllers/UsersController.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    public class UsersController : ControllerBase
+    {
+        private IUserRepository UserRepository { get; }
+
+        public UsersController(IUserRepository userRepository)
+        {
+            UserRepository = userRepository ?? throw new ArgumentNullException(nameof(userRepository));
+        }
+
+        // /api/events
```
Comment (TechGlue) [inline SecretSanta/src/SecretSanta.Api/Controllers/UsersController.cs:21]: Remove or edit the comments to show "// /api/users" Just so anyone looking to call the method from the browser has the correct address.

**Review by TechGlue — COMMENTED** [review]
* Create a UserController in the API projects
    * Implement endpoints for:
        * Retrieving all users ✔
        * Retrieving a single user by id ✔
        * Creating a user ✔
        * Updating a user ✔
        * Deleting a user ✔
    * The UserController should function. ✔
    * Include a .rest file for interacting with the controller (REST Client). ✔
* The IUserRepository interface is provided for you. You will need to provide an implementation of this interface within the SecretSanta.Business project. This implementation can simply store the user within a list, dictionary, array, or similar. ✔
* All test projects should be within a test directory ✔
* Create a unit test project for the SecretSanta.Business project. ✔
    * Unit test all public methods ✔
    * Ensure the project is named with a ".Tests" suffix ✔
    * Ensure test class is named appropriately for the class being tested: ✔
    * Name all tests in the form <MethodBeingTested>_<ConditionBeingTested>_<ExpectedOutcome>(). You can add <NoWarn>CA1707;INTL0003</NoWarn> to remove the naming warnings. ✔
* Create a Unit Test project for the SecretSanta.Api project. ✔
    * Test all controller public methods. It is up to you to determine the best way to provide the controller's dependencies. ❌
    * Ensure the project is named with a ".Tests" suffix ✔
Extra Credit:
* Use a test double library like Moq to provide the controller's dependencies.❌

Everything Looking great. Just had some compiler issues in the SecretSanta.Api.Tests. I’d double check your UserControllerTests.cs and make sure you’re making valid calls to valid methods.

File: SecretSanta/Directory.Build.props [hunk 2]
```diff
@@ -2,4 +2,7 @@
     <ItemGroup>
         <Compile Include="../../SolutionSuppression.cs" />
     </ItemGroup>
+    <PropertyGroup>
```
Comment (Keboo) [inline SecretSanta/Directory.Build.props:5]: Nice job putting this in the Directory.Build.props!

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+namespace SecretSanta.Api.Tests.Controllers
+{
+    [TestClass]
+    public class UsersControllerTests
+    {
+        [TestMethod]
+        //[ExpectedException(typeof(ArgumentNullException))]
+        public void Constructor_WithNullUserRepository_ThrowsAppropriateException()
+        {
+            //Any of the approachs shown here are fine.
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:19]: In general you would pick one of these methods and remove the others.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                return;
+            }
+            Assert.Fail("No exception thrown");
+        }
+
+        [TestMethod]
+        public void Get_WithData_ReturnsUsers()
+        {
+            //Arrange
+            UsersController controller = new(new UserRepository());
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:40]:
```suggestion
            UsersController controller = new(new TestableUserRepository());
```

File: SecretSanta/src/SecretSanta.Business/IUserManager.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using SecretSanta.Data;
+
+namespace SecretSanta.Business
+{
+    public interface IUserManager
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/IUserManager.cs:6]: This appears to be un-used.

**Review by Keboo — COMMENTED** [review]
- Create a UserController in the API projects
  - Implement endpoints for:
    - Retrieving all users ✔
    - Retriving a single user by id ✔
    - Creating a user ✔
    - Updating a user ✔
    - Deleting a user ✔
  - The UserController should function. ✔
  - Include a `.rest` file for interacting with the controller. ✔
- The `IUserRepository` interface is provided for you. You will need to provide an implementation of this interface within the SecretSanta.Business project. This implementation can simply store the user within a list, dictionary, array, or similar. ✔
- All test projects should be within a test directory ✔
- Create a unit test project for the SecretSanta.Business project. ✔
  - Unit test all public methods ✔
  - Ensure the project is named with a ".Tests" suffix ✔
  - Ensure test class is named appropriately for the class being tested: ✔
  - Name all tests in the form `<MethodBeingTested>_<ConditionBeingTested>_<ExpectedOutcome>()`. You can add  `<NoWarn>CA1707;INTL0003</NoWarn>` to remove the naming warnings. ✔
- Create a Unit Test project for the SecretSanta.Api project. ✔
  - Test all controller public methods. It is up to you to determine the best way to provide the controller's dependencies. ✔
  - Ensure the project is named with a ".Tests" suffix ✔

**Pull request closed without merging**